### DISCUSSION TITLE
Remove All Child Widgets

### DISCUSSION THREAD

**Discussion by aaronst** [Textualize/textual#1110] [Q&A] [answered]
I have a container in which I want to set completely different widgets reactively. I've been doing:

```python
def watch_blah(self, new_value: str) -> None:
    for child in self.children:
        child.remove()

    # self.mount(*widgets) based on new_value
```

This works, but I've noticed that it can be a bit slow to remove for more than five or so children. I'm guessing this is because it redraws for every child? Is there a better way to do this? Having a `remove_all()` or `remove_children()` method to remove all or maybe a subset of child widgets in one go would be handy I think.

**Comment by davep** [comment, marked as answer]
One way to do this would be to use the [child combinator](https://textual.textualize.io/guide/CSS/#child-combinator) to get a query of all the children of a particular node, then use the `remove` [loop-free operation](https://textual.textualize.io/guide/queries/#loop-free-operations) with that query.

However... internally it does more or less what you're doing right now and has the same performance hit. I don't think it's been raised as a particular issue just yet, but @willmcgugan has identified that `remove` on a query needs some work to make it faster (this came about last weekend when I ran into the same thing as you, only using what I'm suggesting here, [to remove 1,600 widgets in one go](https://twitter.com/davepdotorg/status/1586831117990199304)).

In short: personally I think `remove` on a query is the way to go and the speed should then improve in a future release of Textual.

**Reply by aaronst** [reply]
Ah yes, I did see the tweets on that. 😆 Thanks for the answer, I'll use `query()` as you suggest.

**Reply by davep** [reply]
Just to update: https://github.com/Textualize/textual/pull/1147 should be incoming with the next release.

**Reply by aaronst** [reply]
Awesome!
